File: briar-android/src/main/java/org/briarproject/briar/android/telegram/TelegramConversationAdapter.java
package org.briarproject.briar.android.telegram;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.briarproject.briar.R;
import org.briarproject.nullsafety.NotNullByDefault;

import androidx.recyclerview.widget.DiffUtil.ItemCallback;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import static org.briarproject.briar.android.util.UiUtils.formatDate;

@NotNullByDefault
final class TelegramConversationAdapter extends
		ListAdapter<TelegramConversationUiMessage,
				TelegramConversationAdapter.MessageViewHolder> {

	TelegramConversationAdapter() {
		super(new MessageCallback());
	}

	@Override
	public MessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(
				R.layout.list_item_telegram_message, parent, false);
		return new MessageViewHolder(view);
	}

	@Override
	public void onBindViewHolder(MessageViewHolder holder, int position) {
		holder.bind(getItem(position));
	}

	static final class MessageViewHolder extends RecyclerView.ViewHolder {

		private final TextView text;
		private final TextView direction;
		private final TextView date;

		MessageViewHolder(View itemView) {
			super(itemView);
			text = itemView.findViewById(R.id.telegramMessageText);
			direction = itemView.findViewById(R.id.telegramMessageDirection);
			date = itemView.findViewById(R.id.telegramMessageDate);
		}

		void bind(TelegramConversationUiMessage item) {
			text.setText(item.getText());
			direction.setText(directionText(item));
			date.setVisibility(dateVisibility(item));
			date.setText(dateText(date.getContext(), item));
		}
	}

	static int dateVisibility(TelegramConversationUiMessage item) {
		return item.getDateMillis() > 0 ? View.VISIBLE : View.GONE;
	}

	static CharSequence dateText(Context context, TelegramConversationUiMessage item) {
		long dateMillis = item.getDateMillis();
		if (dateMillis <= 0) return "";
		return formatDate(context, dateMillis);
	}

	static int directionText(TelegramConversationUiMessage item) {
		return item.isOutgoing() ?
				R.string.telegram_conversation_direction_outgoing :
				R.string.telegram_conversation_direction_incoming;
	}

	private static final class MessageCallback extends
			ItemCallback<TelegramConversationUiMessage> {

		@Override
		public boolean areItemsTheSame(TelegramConversationUiMessage i1,
				TelegramConversationUiMessage i2) {
			return i1.getMessageId() == i2.getMessageId();
		}

		@Override
		public boolean areContentsTheSame(TelegramConversationUiMessage i1,
				TelegramConversationUiMessage i2) {
			return i1.getDateMillis() == i2.getDateMillis() &&
					i1.isOutgoing() == i2.isOutgoing() &&
					i1.getText().equals(i2.getText());
		}
	}
}

File: briar-android/src/main/kotlin/org/briarproject/briar/android/telegram/TelegramConversationAdapter.kt
package org.briarproject.briar.android.telegram

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil.ItemCallback
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import org.briarproject.briar.R
import org.briarproject.briar.android.util.UiUtils.formatDate

class TelegramConversationAdapter :
	ListAdapter<TelegramConversationUiMessage,
		TelegramConversationAdapter.MessageViewHolder>(MessageCallback()) {

	override fun onCreateViewHolder(parent: ViewGroup,
		viewType: Int): MessageViewHolder {
		val view = LayoutInflater.from(parent.context).inflate(
			R.layout.list_item_telegram_message, parent, false)
		return MessageViewHolder(view)
	}

	override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
		holder.bind(getItem(position))
	}

	class MessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		private val text: TextView = itemView.findViewById(R.id.telegramMessageText)
		private val direction: TextView =
			itemView.findViewById(R.id.telegramMessageDirection)
		private val date: TextView = itemView.findViewById(R.id.telegramMessageDate)

		fun bind(item: TelegramConversationUiMessage) {
			text.text = item.text
			direction.setText(directionText(item))
			date.visibility = dateVisibility(item)
			date.text = dateText(date.context, item)
		}
	}

	private class MessageCallback : ItemCallback<TelegramConversationUiMessage>() {

		override fun areItemsTheSame(
			i1: TelegramConversationUiMessage,
			i2: TelegramConversationUiMessage,
		): Boolean = i1.stableId == i2.stableId

		override fun areContentsTheSame(
			i1: TelegramConversationUiMessage,
			i2: TelegramConversationUiMessage,
		): Boolean =
			i1.dateMillis == i2.dateMillis &&
				i1.isOutgoing == i2.isOutgoing &&
				i1.text == i2.text
	}

	companion object {
		@JvmStatic
		fun dateVisibility(item: TelegramConversationUiMessage): Int =
			if (item.dateMillis > 0) View.VISIBLE else View.GONE

		@JvmStatic
		fun dateText(context: Context,
			item: TelegramConversationUiMessage): CharSequence {
			val dateMillis = item.dateMillis
			if (dateMillis <= 0) return ""
			return formatDate(context, dateMillis)
		}

		@JvmStatic
		fun directionText(item: TelegramConversationUiMessage): Int =
			if (item.isOutgoing)
				R.string.telegram_conversation_direction_outgoing
			else R.string.telegram_conversation_direction_incoming
	}
}
